src/string_ptr.rs

```diff
@@ -28,105 +28,153 @@ unsafe impl FromToNativeWasmType for StringPtr {
+    fn alloc(
+        value: &String,
+        env: &Env,
+        memory: &Memory,
+        store: &mut impl AsStoreMut,
+    ) -> anyhow::Result<Box<StringPtr>> {
         let new = export_asr!(fn_new, env);
         let size = i32::try_from(value.len())?;
 
+        // TODO: why 1?
+        // Call __new with parameter: size & class id
+        /*
         let offset = u32::try_from(
-            match new.call(&[Value::I32(size << 1), Value::I32(1)])?.get(0) {
+            match new
+                .call(store, &[Value::I32(size * 2), Value::I32(1)])?
+                .get(0)
+            {
                 Some(val) => match val.i32() {
                     Some(i) => i,
                     _ => anyhow::bail!("Failed to allocate"),
                 },
                 _ => anyhow::bail!("Failed to allocate"),
             },
         )?;
-        write_str(offset, value, env)?;
+        */
+        let offset = u32::try_from(new.call(store, size * 2, 1)?)?;
+        write_str(offset, value, env, memory, store)?;
 
         // pin
         let pin = export_asr!(fn_pin, env);
-        pin.call(&[Value::I32(offset.try_into()?)])?;
+        // pin.call(store, &[Value::I32(offset.try_into()?)])?;
+        pin.call(store, offset.try_into()?)?;
 
         Ok(Box::new(StringPtr::new(offset)))
     }
 
-    fn write(&mut self, value: &String, env: &Env) -> anyhow::Result<Box<StringPtr>> {
-        let memory = match env.memory.get_ref() {
-            Some(mem) => mem,
-            _ => anyhow::bail!("Cannot get memory"),
-        };
-        let prev_size = size(self.offset(), memory)?;
-        let new_size = u32::try_from(value.len())? << 1;
+    fn write(
+        &mut self,
+        value: &String,
+        env: &Env,
+        memory: &Memory,
+        store: &mut impl AsStoreMut,
+    ) -> anyhow::Result<Box<StringPtr>> {
+        let prev_size = size(self, memory, store)?;
+        let new_size = u32::try_from(value.len())?;
+
         if prev_size == new_size {
-            write_str(self.offset(), value, env)?;
+            write_str(self.offset(), value, env, memory, store)?;
             Ok(Box::new(*self))
         } else {
             // unpin old ptr
             let unpin = export_asr!(fn_unpin, env);
-            unpin.call(&[Value::I32(self.offset().try_into()?)])?;
+            // unpin.call(store, &[Value::I32(self.offset().try_into()?)])?;
+            unpin.call(store, self.offset().try_into()?)?;
 
-            // collect
+            // collect (e.g. perform full gc collection)
             let collect = export_asr!(fn_collect, env);
-            collect.call(&[])?;
+            collect.call(store)?;
 
             // alloc with new size
-            StringPtr::alloc(value, env)
+            StringPtr::alloc(value, env, memory, store)
         }
     }
 
-    fn free(self, env: &Env) -> anyhow::Result<()> {
+    fn free(self, _env: &Env, _store: &mut Store) -> anyhow::Result<()> {
+        todo!()
```

Comment:
becareful todo that need to be resolve

Reply:
This is never used and not even unit test. Can we add another issue here to do it properly (e.g. with unit tests)?